### DISCUSSION TITLE
Benchmarks with concurrency

### DISCUSSION THREAD

**Discussion by melugi** [phpbench/phpbench#1026] [General]
First of all just wanted to say thanks for developing this tool, it's been a great help and is always a pleasure to work with, that being said:

Does phpbench support executing benchmarks concurrently to simulate multiple users performing the same activity? It seems like there [was a ticket in the past](https://github.com/phpbench/phpbench/issues/172), and a [PR was made that added concurrency to the runner](https://github.com/phpbench/phpbench/pull/195) but that was closed and seems like it was a WIP.

If it currently isn't supported, are there plans to support it in a future release? Any insights on pitfalls you encountered while working on this previously? It might be worth it on my end to try and come up with a solution in the meantime and push it upstream.

**Comment by dantleech** [comment]
What's your exact use case?

Looks like I decided it was better to use the Promise pattern, I'm not planning on working on this soon, but maybe I cos provide some pointers

**Reply by melugi** [reply]
In my case I'm trying to use phpbench to simulate multiple users interacting with our services, so in a benchmark I include an http client and make web requests using that client. I'd like to be able to add an annotation `@Concurrency(n)` which will then spawn *n* php processes per iteration to 


```
class DocumentBench {
  private $testUsers = [user1, user2, user3...];
  
  private function getUsers()
  {
     return [
      1 => user1,
      2 => user2,
      3 => user3,
  ...
  ];
  }

  // @Iterations(5)
  // @Concurrency(3)
  // @ParamProvider({"getUsers"})
  public function benchInsertDocument($user)
  {
     // This will spawn 3 php processes while the main thread waits on the promises
     // Once all promises are fulfilled we go on to the next iteration
     return new Promise ($client->post(insert_doc_endpoint, payload, user) );
  }
}
```

This is very rough pseudocode though

**Comment by dantleech** [comment]
PHPBench isn't currently the best option for load testing, or at least it would require refactoring to be a good alternative to Gatling or K6.

Basically we want to record the time it takes for the network to respond, not the time it takes to execute some code

Saying that, it might still be worth pursuing this..

**Comment by melugi** [comment]
I figured I might have been trying to use the wrong tool, I looked at K6 and I think I'll end up trying that out as well. Still it would be useful to be able to execute specific parts of the code base while also simulating multiple users, for instance testing load on authentication flow.

If in my example we were just invoking functions would that lessen the effort required to implement this kind of concurrency?

I haven't really explored it but have you found success with pairing phpbench and K6? I understand they serve two different purposes, but it seems to me like having robust K6 coverage would lessen the value of developing phpbench coverage. It still has a lot of value when comparing different implementations. So I'm wondering if you have any insights in using the two together or if you think it's better to stick to one. (Sorry it's a little off topic, no worries if you don't have time to respond 😄)

**Comment by dantleech** [comment]
K6 is more about simulating HTTP users, PHPBench is about testing execution time of PHP code - so for the equivalent of an end to end test K6 is perhaps better, whereas PHPBench is more for unit/integration.
